### ISSUE TITLE
We should work together

### ISSUE THREAD

**Issue by konsumer** [jbeuckm/cordova-plugin-pebble#1]
I am working on an [Android-focused version of the same thing](https://github.com/konsumer/phonegap-pebble) (my project is a few months older.) I was about to start work on the iOS part, and noticed you have done some of that work, but we focused on differnt parts of the API. I feel like it would make our work more useful to a greater audience if we confederated. We could agree on the structure of the JS, and then just combine our work for the 2 targets.

**Comment by jbeuckm** [comment]
Absolutely. I was planning to send you the same email. How should we go
about this?

**Comment by konsumer** [comment]
It looks like we have a lot of things in common. I still need to do the datalog stuff, but I think the rest of the core API is complete (other than `customizeWatchApp`, which I have some ideas about.) I need  to look closer, but we're using different method names. Is this because the underlying SDK looks that way for iOS, or it seemed neater?  I like your method-names better, but I chose mine because they mirror the android SDK, so the examples are easier to port & things are better documented (in upstream plugin.) I am happy to abandon this idea, if we can agree on a useful new API that interfaces with the lower-level stuff.

Your callback style is `[success, failure]` & mine is a single node-style callback: `function(err, return)`. I am fine with yours, and that is actually more like cordova-gateway format, but I think the callbacks should be the last params (which it looks like you do,) so they are optional for things that don't need to wait for completion. Do you want to join as a collaborator on my project, or vice-versa? I don't mind either way, I just want users to have a definitive "this is the one" sort of project to use. I bet if we get it feature-complete, it could even end up in the upstream codebase.

**Comment by konsumer** [comment]
Also, I had an idea to make ngPebble (angular pebble wrapper) as most of the stuff I do these days is angular-based.  I have the start of some demo-ideas [here](https://github.com/konsumer/testpebble/blob/master/www/js/services.js) so maybe this is something that could also benefit your project (if we have the same API.) It  uses `$q` to make callback code much neater.

That project (testpebble) is a fancier "mega-demo"of the stuff I am putting together. It uses ionic, so it's pretty slick. I'd be happy to port over to the new API, if  we can agree on one.

**Comment by konsumer** [comment]
Another issue for mine is that I think `sendDataToPebble`is not quite complete, or at least my "send data to sports app" demo is wrong. I am working on it, but that part might not be 100% functional, yet.

**Comment by konsumer** [comment]
Also, I'm not sure if they are useful on the iOS-end, but I needed the constants to make the built-in apps work, so that might be another thing we should combine.  I wrote a [simple lil script](https://github.com/konsumer/phonegap-pebble/blob/master/parse_constants.js) to grab them from the java-source.

**Comment by konsumer** [comment]
Alternately, maybe we could both make higher-level functions that just do the update for sports & golf, and then we don't need all the constants.

**Comment by jbeuckm** [comment]
Your plugin is more developed and follows the SDK more closely. I think I should bend mine toward your api implementation and eventually make a pull request. Before that, I need to get appMessage to work consistently. It is failing sometimes. I will be taking a look at how you implemented messaging and please look at mine if you have time.

**Comment by nicola** [comment]
+1 please :)

**Comment by konsumer** [comment]
I've made some progress getting my callbacks more like yours (success, fail) and making sure that event callbacks (connect, disconnect, ack, nack, message, etc) are firing, which now they fire on the document, so they act more like web-ish events. It seems like all is working ok, but I think I must be missing something about how message callbacks work because I am also having trouble with these. I am getting a lot of NACK & dropped packets on user-initiated messages. You can see my [current demo in the plugin repo](https://github.com/konsumer/phonegap-pebble/blob/master/demo/phone/www/index.html) whenever the user clicks the "send message" link, it sends, but most of the time it drops if they do it very fast. I wonder if messages need to be queued & ACK/NACK needs to be checked on all. If your plugin has worked around this issue, I'd love to hear about it.

**Comment by chrleon** [comment]
How is this for usage in an app? Should I use jbeuckm's or konsumer's plugin? I'm just sending automated updated to the watch, like at fixed times.

**Comment by jameslegue** [comment]
Any update on the status of merging the two repos together? I am very interested in using it both for iOS and Android.

@konsumer did you get anywhere with the ngPebble? Also very interested to use in Ionic :)

**Comment by konsumer** [comment]
Mine is good for android, @jbeuckm's is good for iOS. I haven't heard anything about the integration work progressing and have been too busy to do it myself. Eventually, I think the API will look like mine, as It's more complete, so it might make sense to get your app working on android first, as the code wont change once they are integrated. I welcome PRs, and can help a bit, but I am pretty busy.

**Comment by konsumer** [comment]
As for ng, I don't use angular too much anymore, but it should be trivial to use with angular, in whatever way works for your app structure. Specific angular bindings isn't really needed, just [bind](https://github.com/konsumer/pebble-comtest/blob/master/app/www/index.html#L40-L45) to pebble message events, or [send messages](https://github.com/konsumer/pebble-comtest/blob/master/app/www/index.html#L54) to your pebble.

**Comment by konsumer** [comment]
You guys should check out [this](https://github.com/konsumer/pebble-comtest). I used it while developing, and it gives a complete example of how to boilerplate your project. Both of our projects seem to be a lil unreliable with ACK/NACK, so just leave that out (sending and receiving works fine, just lower-level messages seem flakey.) I think it might be due to the lower-level API already doing this stuff in the background, but I'm not sure.
